src/Processors/Formats/RowInputFormatWithNamesAndTypes.cpp

```diff
@@ -52,53 +72,125 @@ void RowInputFormatWithNamesAndTypes::readPrefix()
+                ErrorCodes::INCORRECT_DATA,
+                "The number of data types differs from the number of column names in input data");
+
+        /// Check that types from input matches types from header.
+        for (size_t i = 0; i < type_names.size(); ++i)
         {
-            auto types = format_reader->readTypes();
-            if (types.size() != column_mapping->column_indexes_for_input_fields.size())
+            if (column_mapping->column_indexes_for_input_fields[i] &&
+                data_types[*column_mapping->column_indexes_for_input_fields[i]]->getName() != type_names[i])
+            {
                 throw Exception(
                     ErrorCodes::INCORRECT_DATA,
-                    "The number of data types differs from the number of column names in input data");
-
-            /// Check that types from input matches types from header.
-            for (size_t i = 0; i < types.size(); ++i)
-            {
-                if (column_mapping->column_indexes_for_input_fields[i] &&
-                    data_types[*column_mapping->column_indexes_for_input_fields[i]]->getName() != types[i])
-                {
-                    throw Exception(
-                        ErrorCodes::INCORRECT_DATA,
-                        "Type of '{}' must be {}, not {}",
-                        getPort().getHeader().getByPosition(*column_mapping->column_indexes_for_input_fields[i]).name,
-                        data_types[*column_mapping->column_indexes_for_input_fields[i]]->getName(), types[i]);
-                }
+                    "Type of '{}' must be {}, not {}",
+                    getPort().getHeader().getByPosition(*column_mapping->column_indexes_for_input_fields[i]).name,
+                    data_types[*column_mapping->column_indexes_for_input_fields[i]]->getName(), type_names[i]);
             }
         }
-        else
-            format_reader->skipTypes();
     }
 }
 
+void RowInputFormatWithNamesAndTypes::tryDetectHeader(std::vector<String> & column_names_out, std::vector<String> & type_names_out)
+{
+    auto & read_buf = getReadBuffer();
+    PeekableReadBuffer * peekable_buf = dynamic_cast<PeekableReadBuffer *>(&read_buf);
+    if (!peekable_buf)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Header detection is supported only for formats that use PeekableReadBuffer");
+
+    /// Empty data.
+    if (unlikely(format_reader->checkForSuffix()))
+    {
+        end_of_stream = true;
+        return;
+    }
+
+    /// Make a checkpoint before reading the first row.
+    peekable_buf->setCheckpoint();
+    auto first_row_values = format_reader->readRowForHeaderDetection();
+
+    /// To understand if the first row is a header with column names, we check
+    /// that all values from this row is a subset of column names from provided header.
+    auto column_names = getPort().getHeader().getNames();
+    std::unordered_set<String> column_names_set(column_names.begin(), column_names.end());
```

Comment:
```suggestion
    std::unordered_set<std::string_view> column_names_set(column_names.begin(), column_names.end());
```
not sure if it would work

Reply:
It works, thanks!